### PR TITLE
POC: Use LRU for data_dict

### PR DESCRIPTION
**Pull request by RahulKushwaha**
Testing waters if we convert data_dict to lru

### CODE REVIEW

File: include/dict0types.h [hunk 1]
```diff
@@ -530,6 +534,50 @@ struct dict_table_t {
   IF_DEBUG(ulint magic_n;)
 };
 
+class dict_table_lru {
+ private:
+  using table_list_t = std::list<dict_table_t *>;
```
Comment (sunbains) [inline include/dict0types.h:539]: You are better off naming this Tables and the Iterator  as Iter. It shouldn't be important what the underlying data structure is.

File: include/dict0types.h [hunk 2]
```diff
@@ -530,6 +534,50 @@ struct dict_table_t {
...
+  using list_itr_t = table_list_t::iterator;
+
+  struct cache_entry {
+    list_itr_t list_itr;
```
Comment (sunbains) [inline include/dict0types.h:543]: Prefix struct/class members with m_ easier to distinguish in the code.

File: include/dict0types.h [hunk 3]
```diff
@@ -530,6 +534,50 @@ struct dict_table_t {
   IF_DEBUG(ulint magic_n;)
 };
 
+class dict_table_lru {
+ private:
+  using table_list_t = std::list<dict_table_t *>;
+  using list_itr_t = table_list_t::iterator;
+
+  struct cache_entry {
```
Comment (sunbains) [inline include/dict0types.h:542]: For classes we need to move away from the current naming and use upper case first letter and snake case.

File: dict/dict0dict.cc [hunk 4]
```diff
@@ -548,39 +543,35 @@ void dict_table_add_to_cache(dict_table_t *table, mem_heap_t *heap) {
...
-    dict_table_t *table2;
-    HASH_SEARCH(
-      name_hash, dict_sys->table_hash, fold, dict_table_t *, table2, ut_ad(table2->cached), strcmp(table2->name, table->name) == 0
-    );
-    ut_a(table2 == nullptr);
+  if (!skip_add_to_cache) {
+
+    /* Look for a table with the same name: error if such exists */
+    {
+      auto contains = dict_sys->table_lookup->contains(table->name);
```
Comment (sunbains) [inline dict/dict0dict.cc:550]: Better name for contains will be found

File: dict/dict0dict.cc [hunk 5]
```diff
@@ -548,39 +543,35 @@ void dict_table_add_to_cache(dict_table_t *table, mem_heap_t *heap) {
...
+      ut_a(!contains);
 
 #ifdef UNIV_DEBUG
-    /* Look for the same table pointer with a different name */
-    HASH_SEARCH_ALL(name_hash, dict_sys->table_hash, dict_table_t *, table2, ut_ad(table2->cached), table2 == table);
-    ut_ad(table2 == nullptr);
+      /* Look for the same table pointer with a different name */
```
Comment (sunbains) [inline dict/dict0dict.cc:554]: Is this hash used?

File: dict/dict0dict.cc [hunk 6]
```diff
@@ -857,9 +844,14 @@ void dict_table_remove_from_cache(dict_table_t *table) {
...
 
-  /* Remove table from the hash tables of tables */
-  HASH_DELETE(dict_table_t, name_hash, dict_sys->table_hash, ut_fold_string(table->name), table);
-  HASH_DELETE(dict_table_t, id_hash, dict_sys->table_id_hash, ut_uint64_fold(table->id), table);
+  if (!skip_remove_from_lru) {
+    /* Remove table from lru */
+    ut_a(dict_sys->table_lookup->contains(table->name));
+    ut_a(dict_sys->table_lookup->contains(table->id));
+
+    dict_sys->table_lookup->erase(table->id);
```
Comment (sunbains) [inline dict/dict0dict.cc:852]: Use the lower_bound and iterator to remove the element. IIRC, it makes it an O(1) operation. You already have  lookup/contains in a runtime assertion.

File: dict/dict0dict.cc [hunk 7]
```diff
@@ -3885,7 +3819,208 @@ void dict_close() {
...
   dict_sys = nullptr;
 
-  for (ulint i = 0; i < DICT_INDEX_STAT_MUTEX_SIZE; i++) {
-    mutex_free(&dict_index_stat_mutex[i]);
+  for (auto &i : dict_index_stat_mutex) {
+    mutex_free(&i);
   }
 }
+
+dict_table_lru::dict_table_lru(std::int32_t capacity, std::function<dict_table_t *(std::uint64_t)> table_fetcher)
```
Comment (sunbains) [inline dict/dict0dict.cc:3827]: My preference is to create a dict namespace and name the class Table_LRU within that.

File: dict/dict0dict.cc [hunk 8]
```diff
@@ -3885,7 +3819,208 @@ void dict_close() {
...
+    : m_capacity{capacity}, m_size{0}, m_table_fetcher{std::move(table_fetcher)}, m_tables{}, m_name_lookup{}, m_id_lookup{} {}
+
+dict_table_t *dict_table_lru::get(std::string_view name) {
+  if (auto itr = m_name_lookup.find(name); itr != m_name_lookup.end()) {
```
Comment (sunbains) [inline dict/dict0dict.cc:3831]: Most books tend to use it for iterator, I'm partial to shorter names.

File: dict/dict0dict.cc [hunk 9]
```diff
@@ -3885,7 +3819,208 @@ void dict_close() {
...
+    ut_a(m_id_lookup[id] == m_name_lookup[itr->second->name]);
+    move_to_front(itr->second);
+    return (*itr->second->list_itr);
+  }
+
+  return nullptr;
+}
+
+bool dict_table_lru::contains(std::uint64_t id) {
+  ib_logger(ib_stream, "contains check table: %lu\n", id);
```
Comment (sunbains) [inline dict/dict0dict.cc:3851]: When writing new error logging messages use log_err(), log_warn(), log_info() etc. I want to get rid of this ib_stream mechanism.

File: dict/dict0dict.cc [hunk 10]
```diff
@@ -3885,7 +3819,208 @@ void dict_close() {
...
+  return m_id_lookup.contains(id);
+}
+
+bool dict_table_lru::contains(std::string_view name) {
```
Comment (sunbains) [inline dict/dict0dict.cc:3855]: Remember to tag the functions as const and [[nodiscard]] and noexcept where required.

File: dict/dict0dict.cc [hunk 11]
```diff
@@ -3885,7 +3819,208 @@ void dict_close() {
...
+    itr = m_tables.begin();
+  }
+
+  ut_a(itr != m_tables.end());
+
+  entry->evicted = false;
+  entry->list_itr = itr;
+}
+
+void dict_table_lru::evict() {
```
Comment (sunbains) [inline dict/dict0dict.cc:3972]: There are at least two problems:

1. Foreign key dependencies, you can't remove just one of the tables.
2. Some state is not persistent e.g., auto-increment. We need to preserve that even if we evict the table.

Write a unit-test to check for the above two.

Comment (RahulKushwaha) [reply dict/dict0dict.cc:3972]: oh ya. Just realized that the state will not persistent correctly. It must be recoverable but not consistent with file block. I will play with this more.

**Comment by RahulKushwaha** [comment]
Closed for now. I will pick it up a little later.

**Pull request closed without merging**